gcsfs/core.py

```diff
@@ -1060,13 +1061,32 @@ async def _ls(
             for entry in await self._list_objects(
                 path, prefix=prefix, versions=versions, **kwargs
             ):
+                if entry["size"] == 0 and entry["name"].endswith("/"):
+                    entry = {
+                        "bucket": entry["bucket"],
+                        "name": path.rstrip("/"),
+                        "size": 0,
+                        "storageClass": "DIRECTORY",
+                        "type": "directory",
+                    }
+                    if entry in out:
```

Comment:
Might this comparison of a dict to all the dicts so far be expensive?

Reply:
Yes, that is expensive. The check is necessary, directories should not be listed twice when there exists both a placeholder file and a file under that prefix. Directories is not hashable. I could make a temporary set of directory paths to check against?

Reply:
> I could make a temporary set of directory paths to check against?

Yes, I'm afraid we probably need that